sql/planner/hash_set.go

```diff
@@ -0,0 +1,65 @@
+package planner
+
+import (
+	"hash"
+	"hash/maphash"
+
+	"github.com/genjidb/genji/document"
+	"github.com/genjidb/genji/key"
+)
+
+type documentHashSet struct {
+	hash hash.Hash64
+	set  map[uint64]struct{}
+}
+
+func newDocumentHashSet(hash hash.Hash64) *documentHashSet {
+	if hash == nil {
+		hash = &maphash.Hash{}
+	}
+
+	return &documentHashSet{
+		hash: hash,
+		set:  map[uint64]struct{}{},
+	}
+}
+
+func (s documentHashSet) generateKey(d document.Document) (uint64, error) {
+	defer s.hash.Reset()
+
+	err := d.Iterate(func(field string, value document.Value) error {
+		var buf []byte
+		buf, err := key.AppendValue(buf, value)
+		if err != nil {
+			return err
+		}
+
+		_, err = s.hash.Write(buf)
+		if err != nil {
+			return err
+		}
+
+		buf = buf[0:]
+		return nil
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	return s.hash.Sum64(), nil
+}
```

Comment:
This function relies on `Iterate` function implementation, so if it does not guarantee order of field iteration (for example `map`-based) then hash will be incorrect.
For now, Genji does not have such implementation, but it can be added later.
I think `document.Value`/`document.Document` should have some `Hash` function to use it in hash table.

Reply:
We should not rely on the implicit behavior of Document implementations as the `Document` interface doesn't guarantee the order of iteration.
Instead, we must use the [document.Fields()](https://github.com/genjidb/genji/blob/c05715e76a2436868192fc3ac47dc9e8c18e3437/document/document.go#L60) function, which gets the document keys and sorts them, and use `GetByField` for each key.

Reply:
`pk()`/`literal` tests was failing, when I used `documentMask.GetByField`, it tries to get a value from masked document, but literals and functions values are not stored in document.
So, I added  `IterateInOrder` function.

Reply:
It appears to be a bug, they should be stored in `DocumentMask`:

```go
genji> create table foo;
genji> insert into foo(a) values (1), (2), (3);
genji> select pk() from foo;
{
  "pk()": 1
}
{
  "pk()": 2
}
{
  "pk()": 3
}
```